Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,258 @@
+    (fun (i : ℕ) ↦ (filter (fun t ↦ card t.fst = i) (pairs σ k))) pdisj
+  have disj_equiv : disjiUnion (range k) (fun i ↦ filter (fun t ↦ card t.fst = i) (pairs σ k))
+      pdisj = filter (fun t ↦ card t.fst < k) (pairs σ k) := by
+    ext a
+    rw [mem_disjiUnion, mem_filter]
+    refine' ⟨_, fun haf ↦ ⟨card a.fst, mem_range.mpr haf.right, _⟩⟩
+    · rintro ⟨n, hnk, ha⟩
+      have hnk' : n ≤ k := by apply le_of_lt; aesop
+      aesop
+    · simp_all [mem_filter]
+  simp only [← hdisj, disj_equiv]
+
+private theorem disjoint_filter_pairs_lt_filter_pairs_eq (k : ℕ) :
+    Disjoint (filter (fun t ↦ card t.fst < k) (pairs σ k))
+    (filter (fun t ↦ card t.fst = k) (pairs σ k)) := by
+  rw [disjoint_filter]
+  exact fun _ _ h1 h2 ↦ lt_irrefl _ (h2.symm.subst h1)
+
+private theorem disjUnion_filter_pairs_eq_pairs (k : ℕ) :
+    disjUnion (filter (fun t ↦ card t.fst < k) (pairs σ k))
+    (filter (fun t ↦ card t.fst = k) (pairs σ k)) (disjoint_filter_pairs_lt_filter_pairs_eq σ k) =
+    pairs σ k := by
+  simp only [disjUnion_eq_union, Finset.ext_iff, pairs, filter_filter, mem_filter]
+  intro a
+  rw [← filter_or, mem_filter]
+  refine' ⟨fun ha ↦ by tauto, fun ha ↦ _⟩
+  have hacard := le_iff_lt_or_eq.mp ha.2.1
+  tauto
+
+private theorem esymm_summand_to_weight (k : ℕ) (A : Finset σ) (h : A ∈ powersetLen k univ) :
+    ∑ j in A, weight σ R k (A, j) = k * (-1) ^ k * (∏ i in A, X i : MvPolynomial σ R) := by
+  simp [weight, mem_powerset_len_univ_iff.mp h, mul_assoc]
+
+private theorem esymm_to_weight (k : ℕ) : k * esymm σ R k =
+    (-1) ^ k * ∑ t in filter (fun t ↦ card t.fst = k) (pairs σ k), weight σ R k t := by
+  rw [esymm, sum_filter_pairs_eq_sum_powersetLen_sum σ R k (fun t ↦ weight σ R k t),
+    sum_congr rfl (esymm_summand_to_weight σ R k), mul_comm (k : MvPolynomial σ R) ((-1) ^ k),
+    ← mul_sum, ← mul_assoc, ← mul_assoc, ← pow_add, Even.neg_one_pow ⟨k, rfl⟩, one_mul]
+
+private theorem esymm_mul_psum_summand_to_weight (k i : ℕ) :
+    ∑ A in powersetLen i univ, ∑ j, weight σ R k (A, j) =
+    (-1) ^ i * esymm σ R i * psum σ R (k - i) := by
+  simp only [esymm, psum_def, weight, ← mul_assoc, mul_sum]
+  rw [sum_comm]
+  refine' sum_congr rfl fun x _ ↦ _
+  rw [sum_mul]
+  refine' sum_congr rfl fun s hs ↦ _
+  rw [mem_powerset_len_univ_iff.mp hs]
+
+private theorem esymm_mul_psum_to_weight (k : ℕ) :
+    ∑ i in range k, (-1) ^ i * esymm σ R i * psum σ R (k - i) =
+    ∑ t in filter (fun t ↦ card t.fst < k) (pairs σ k), weight σ R k t := by
+  rw [← sum_congr rfl (fun i _ ↦ esymm_mul_psum_summand_to_weight σ R k i),
+    sum_filter_pairs_eq_sum_range_powersetLen_sum σ R k]
+
+/-- **Newton's identities** give a recurrence relation for the kth elementary symmetric polynomial
+in terms of lower degree elementary symmetric polynomials and power sums. -/
+theorem esymm_recurrence (k : ℕ) : (-1) ^ k * (k * esymm σ R k) +
+    ∑ i in range k, (-1) ^ i * esymm σ R i * psum σ R (k - i) = 0 := by
```

Comment:
Note that this is a theorem and not a definition so we should try to name it based on the type (so "recurrence" has no place here or below). Also, I think the following is a more useful statement:
```suggestion
theorem mul_esymm_eq_sum (k : ℕ) :
    k * esymm σ R k = (-1)^(k+1) * ∑ i in range k, (-1) ^ i * esymm σ R i * psum σ R (k - i) := by
```

Reply:
Done!